File: ios/LLMHub/Sources/LLMHub/MimoBot/Kokoro/KokoroVocab.swift
import Foundation

/// Canonical Kokoro phoneme vocabulary. Mirrors
/// `hexgrad/kokoro/__init__.py: SYMBOLS = [PAD] + PUNCT + LETTERS + LETTERS_IPA`
/// — exact order is required, token IDs are positional.
enum KokoroVocab {

    static let pad = "$"

    static let symbols: [String] = {
        var s: [String] = []
        s.append(pad)
        // 16 punctuation
        s.append(contentsOf: [";", ":", ",", ".", "!", "?", "¡", "¿", "—", "…", "\"", "«", "»", "“", "”", " "])
        // 52 ASCII letters
        s.append(contentsOf: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".map { String($0) })
        // 109 IPA — must match LETTERS_IPA in hexgrad/kokoro
        s.append(contentsOf: [
            "ɑ","ɐ","ɒ","æ","b","ɓ","ʙ","β","c","ɕ","ç","ɗ","ɖ","ð","ʤ","ə",
            "ɘ","ɚ","ɛ","ɜ","ɝ","e","ɞ","ɟ","ʄ","ɡ","ɠ","ɢ","ʛ","h","ɦ","ɧ",
            "ħ","ɥ","ʜ","i","ɪ","ɨ","ɪ̆","j","ʝ","ɟ","ʄ","k","ɭ","l","ɫ","ɮ",
            "ɬ","ʟ","m","ɱ","n","ɳ","ɲ","ŋ","ɴ","o","ɵ","ø","œ","ɶ","ɔ","ɸ",
            "p","ɸ","q","ʁ","ɽ","ɾ","r","ɹ","ɻ","ʀ","ʀ̥","ɽ","s","ʂ","ɕ","ʃ",
            "ʈ","t","ʈ","ʇ","θ","u","ʉ","ʊ","ʋ","v","ⱱ","ʌ","w","ʍ","x","χ",
            "y","ʏ","ʎ","z","ʑ","ʐ","ʒ","ʔ","ʕ","ʡ","ʢ"," ","ǀ","ǁ","ǂ","ǃ",
            "ˈ","ˌ","ː","ˑ","ʼ","ʴ","ʰ","ʱ","ʲ","ʷ","ˠ","ˤ","˞","↓","↑","→",
            "↗","↘","'","̩","'","ᵻ"
        ])
        return s
    }()

    static let idBySymbol: [String: Int] = {
        var m: [String: Int] = [:]
        for (i, s) in symbols.enumerated() { m[s] = i }
        return m
    }()

    /// Tokenise an IPA phoneme string as Unicode grapheme clusters so that
    /// e.g. "ɪ̆" stays a single token.
    static func tokenize(_ ipa: String) -> [Int] {
        var out: [Int] = []
        for cluster in ipa.unicodeScalars.grouped() {
            let s = String(String.UnicodeScalarView(cluster))
            if let id = idBySymbol[s] {
                out.append(id)
            } else if let first = cluster.first, let id = idBySymbol[String(first)] {
                out.append(id)
            }
        }
        return out
    }
}

/// Group code points into clusters: a base character followed by any combining
/// marks. Mirrors the Android implementation that merges NON_SPACING_MARK
/// (combining) code points onto the preceding base.
private extension String.UnicodeScalarView {
    func grouped() -> [[Unicode.Scalar]] {
        var clusters: [[Unicode.Scalar]] = []
        for scalar in self {
            let cat = Unicode.Scalar(scalar.value)?.properties.generalCategory
            let isCombining = cat == .nonspacingMark || cat == .spacingMark || cat == .enclosingMark
            if isCombining, !clusters.isEmpty {
                clusters[clusters.count - 1].append(scalar)
            } else {
                clusters.append([scalar])
            }
        }
        return clusters
    }
}

File: android/app/src/main/java/com/llmhub/llmhub/mimobot/speech/kokoro/KokoroVocab.kt
package com.llmhub.llmhub.mimobot.speech.kokoro

/**
 * Canonical Kokoro phoneme vocabulary.
 *
 * Token IDs must match the export at
 *   https://huggingface.co/onnx-community/Kokoro-82M-ONNX
 * which itself follows hexgrad/kokoro/__init__.py: SYMBOLS = [PAD] + PUNCT + LETTERS + LETTERS_IPA.
 *
 * Each entry is a single grapheme (one IPA "character" — some are multi-codepoint
 * like the i-with-breve `ɪ̆`). When tokenising we walk the input as Unicode
 * grapheme clusters, not raw chars, to keep these aligned.
 */
object KokoroVocab {

    const val PAD = "$"

    // The ordered list of symbols. Index = token ID.
    val SYMBOLS: List<String> = buildList {
        // 0 — pad
        add(PAD)
        // 1..16 — punctuation (16 entries)
        addAll(listOf(";", ":", ",", ".", "!", "?", "¡", "¿", "—", "…", "\"", "«", "»", "“", "”", " "))
        // 17..68 — uppercase + lowercase ASCII letters (52)
        addAll("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".map { it.toString() })
        // 69+ — IPA letters / diacritics (109 entries)
        // NOTE: order MUST match hexgrad/kokoro/__init__.py LETTERS_IPA exactly.
        addAll(listOf(
            "ɑ","ɐ","ɒ","æ","b","ɓ","ʙ","β","c","ɕ","ç","ɗ","ɖ","ð","ʤ","ə",
            "ɘ","ɚ","ɛ","ɜ","ɝ","e","ɞ","ɟ","ʄ","ɡ","ɠ","ɢ","ʛ","h","ɦ","ɧ",
            "ħ","ɥ","ʜ","i","ɪ","ɨ","ɪ̆","j","ʝ","ɟ","ʄ","k","ɭ","l","ɫ","ɮ",
            "ɬ","ʟ","m","ɱ","n","ɳ","ɲ","ŋ","ɴ","o","ɵ","ø","œ","ɶ","ɔ","ɸ",
            "p","ɸ","q","ʁ","ɽ","ɾ","r","ɹ","ɻ","ʀ","ʀ̥","ɽ","s","ʂ","ɕ","ʃ",
            "ʈ","t","ʈ","ʇ","θ","u","ʉ","ʊ","ʋ","v","ⱱ","ʌ","w","ʍ","x","χ",
            "y","ʏ","ʎ","z","ʑ","ʐ","ʒ","ʔ","ʕ","ʡ","ʢ"," ","ǀ","ǁ","ǂ","ǃ",
            "ˈ","ˌ","ː","ˑ","ʼ","ʴ","ʰ","ʱ","ʲ","ʷ","ˠ","ˤ","˞","↓","↑","→",
            "↗","↘","'","̩","'","ᵻ"
        ))
    }

    private val ID_BY_SYMBOL: Map<String, Int> = SYMBOLS.withIndex().associate { (i, s) -> s to i }

    fun idOf(symbol: String): Int? = ID_BY_SYMBOL[symbol]

    /** Tokenise an IPA phoneme string as Unicode grapheme clusters. */
    fun tokenise(ipa: String): IntArray {
        if (ipa.isEmpty()) return IntArray(0)
        val out = ArrayList<Int>(ipa.length)
        // Walk by code-point, then merge combining marks into the previous cluster.
        var i = 0
        while (i < ipa.length) {
            val cp = ipa.codePointAt(i)
            val baseLen = Character.charCount(cp)
            // Greedily attach combining marks (e.g. "̩" U+0329, "̥" U+0325).
            var j = i + baseLen
            while (j < ipa.length) {
                val cp2 = ipa.codePointAt(j)
                if (Character.getType(cp2) != Character.NON_SPACING_MARK.toInt()) break
                j += Character.charCount(cp2)
            }
            val cluster = ipa.substring(i, j)
            ID_BY_SYMBOL[cluster]?.let(out::add)
                ?: ID_BY_SYMBOL[ipa.substring(i, i + baseLen)]?.let(out::add)
            i = j
        }
        return out.toIntArray()
    }
}
